### PR TITLE
Map memory serialization fixes, tests, and performance

### PR DESCRIPTION
**Pull request by jbytheway**
#### Summary
SUMMARY: Bugfixes "Map memory no longer reversed through save/load cycle"

#### Purpose of change
I started this branch because I wanted to improve `map_memory` performance on `_GLIBCXX_DEBUG` builds.  Along the way I found and fixed a rather important bug.

#### Describe the solution
* The order of map memory was being reversed through every save/load cycle.  So, shortly after loading a game you might suddenly forget all the place you had most recently visited.  I fixed it by placing newly memorized tiles at the end of the cache list rather than the beginning.
* I added a test to catch this particular bug and a few other obvious `map_memory` tests.
* I factored out the two LRU cache implementations into a single class template.

#### Performance
The performance question turned out to be quite complicated.  The important change is that for the common case of moving an existing entry in the list I've replaced `erase`+`insert` with `splice`.  This is a good idea for release or debug builds, since it means that no memory allocations are required.

However, I was really trying to improve the performance of the `_GLIBCXX_DEBUG` build.  It turns out that both `erase` and `splice` are O(n) operations for the debug list, so both options are very slow (and this is slow enough to be a problem in practice, which is why I was looking into it).  However, the difference is that `erase` must be slow, whereas `splice` can be fast.  I've patched my local libstdc++ to make it fast, and the difference is dramatic.  For details see the [issue I've opened](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=87872) on the gcc Bugzilla.

If anyone else wants a similar performance boost they can apply the following patch to their standard library headers, or hope that my suggestion in the above issue is implemented, in which case this might eventually get faster in gcc 9.
```patch
--- debug/list
+++ debug/list-new
@@ -582,7 +582,9 @@
 
        // _GLIBCXX_RESOLVE_LIB_DEFECTS
        // 250. splicing invalidates iterators
-       this->_M_transfer_from_if(__x, _Equal(__i.base()));
+       if (this != &__x) {
+         this->_M_transfer_from_if(__x, _Equal(__i.base()));
+       }
        _Base::splice(__position.base(), _GLIBCXX_MOVE(__x._M_base()),
                      __i.base());
       }
```

#### Describe alternatives you've considered
~~Note that the new `lru_cache::get` returns `T{}`, which relies on the fact that a value-initialized T can be distinguished from any 'correct' value.  This is how the old code worked, but in the generic context it might be argued that returning an `optional` would be more appropriate.  I have chosen not to do that, so as to minimize the changes (and perhaps to get marginally better performance).~~  Changed to work around older gcc warnings.

### CODE REVIEW

File: src/map_memory.cpp [hunk 1]
```diff
@@ -2,62 +2,78 @@
...
+}
+
+template<typename T>
+const std::list<typename lru_cache<T>::Pair> &lru_cache<T>::list() const
+{
+    return ordered_list;
+}
+
+template class lru_cache<memorized_terrain_tile>;
+template class lru_cache<long>;
```
[the remark this answers is not in the document]
Comment (kevingranade) [reply src/map_memory.cpp]: See https://en.cppreference.com/w/cpp/language/class_template#Explicit_instantiation
tl;dr makes template declaration and definition more convenient and faster to compile.

**Comment by kevingranade** [comment]
GCC 4.9 and under errors on the initializer statement for non-present element lookup.
```
g++-4.8  -DRELEASE -DGIT_VERSION -DBACKTRACE -DLOCALIZE -ffast-math -Os  -Wall -Wextra  -g  -fsigned-char -Werror -std=c++11 -MMD -MP -I/usr/include/ncursesw   -c src/map_memory.cpp -o obj/map_memory.o
src/map_memory.cpp: In instantiation of ‘T lru_cache<T>::get(const tripoint&) const [with T = memorized_terrain_tile]’:
src/map_memory.cpp:57:16:   required from here
src/map_memory.cpp:12:14: error: missing initializer for member ‘memorized_terrain_tile::tile’ [-Werror=missing-field-initializers]
     return T{}; // Value or aggregate initialized (in particular, integers are zero-initialized)
              ^
src/map_memory.cpp:12:14: error: missing initializer for member ‘memorized_terrain_tile::subtile’ [-Werror=missing-field-initializers]
src/map_memory.cpp:12:14: error: missing initializer for member ‘memorized_terrain_tile::rotation’ [-Werror=missing-field-initializers]
```

File: src/map_memory.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (jbytheway) [reply src/map_memory.cpp]: [Explicit instantiation](https://en.cppreference.com/w/cpp/language/class_template#Explicit_instantiation), to allow the implementation to be in the cpp file.  Which vastly speeds up the process of tweaking/debugging this code.
